### PR TITLE
do not remove local directory when out is local

### PR DESCRIPTION
**Pull request by XiaohanZhangCMU**
## Description of changes:

When user forgets to set keep_local and out=local_path, all shards and index files will be removed in the end. Even though the code follows the correct logic, it's bad use experience. 

Changed logic: when out is a local path, only when the user specifically set keep_local=False, we remove the local, otherwise, keep the locals. 

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [ ] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [ ] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [ ] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

**Review by karan6181 — APPROVED** [review]

File: tests/base/converters/test_dataframe_to_mds.py [hunk 1]
```diff
@@ -98,9 +98,9 @@ def test_end_to_end_conversion_local_nocolumns(self, dataframe: Any, keep_local:
                             nsamples += shards[0]['samples']
                 assert nsamples == sum([a['samples'] for a in mgi['shards']])
             if not keep_local:
-                assert not os.path.exists(os.path.join(
-                    out,
-                    'index.json')), 'merged index.json is found even through keep_local = False'
+                assert os.path.exists(
+                    os.path.join(out, 'index.json')
+                ), 'merged index.json is not found keep_local = False but no remote exist'
```
Comment (snarayan21) [inline tests/base/converters/test_dataframe_to_mds.py:103]:
```suggestion
                ), 'merged index.json was not found. keep_local is False but no remote exists.'
```

File: tests/base/converters/test_dataframe_to_mds.py [hunk 2]
```diff
@@ -178,8 +178,9 @@ def test_end_to_end_conversion_local(self, dataframe: Any, keep_local: bool, mer
                             nsamples += shards[0]['samples']
                 assert nsamples == sum([a['samples'] for a in mgi['shards']])
             else:
-                assert not os.path.exists(os.path.join(
-                    out, 'index.json')), 'merged index.json is found even keep_local=False'
+                assert os.path.exists(
+                    os.path.join(out, 'index.json')
+                ), 'merged index.json is not found when keep_local=False but no remote part exists'
```
Comment (snarayan21) [inline tests/base/converters/test_dataframe_to_mds.py:183]:
```suggestion
                ), 'merged index.json was not found. keep_local is False, but no remote part exists.'
```

File: tests/base/converters/test_dataframe_to_mds.py (hunk 1 above)
Comment (snarayan21) [reply tests/base/converters/test_dataframe_to_mds.py:103]: nit

File: tests/base/converters/test_dataframe_to_mds.py (hunk 2 above)
Comment (snarayan21) [reply tests/base/converters/test_dataframe_to_mds.py:183]: nit

**Review by snarayan21 — APPROVED** [review]

File: tests/base/converters/test_dataframe_to_mds.py (hunk 1 above)
Comment (xiaohanzhan-db) [reply tests/base/converters/test_dataframe_to_mds.py:103]: Thanks. Done.

File: tests/base/converters/test_dataframe_to_mds.py (hunk 2 above)
Comment (xiaohanzhan-db) [reply tests/base/converters/test_dataframe_to_mds.py:183]: Thanks, done.

**Pull request merged**